Show static participant list

cheukchuen opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
* Show static participant list through the people button on the top banner message

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What to Check
Verify that the following are valid
* Single local participant should show one row
* Multiple participants should grow the list
* List stops at mid screen, which then allows for scrolling
* Changes to the participants (join, leave, mute, unmute) should update the list by dismissing and reopen the list
* Bottom drawer list should function the same for selecting audio devices

cheukchuen committed:
Show static participant list

Gemakk reviewed (dismissed):
Tested with many remote participants.
Just one minor suggestion: Expand the Label maximum size to use up all available space before the checkmark for long names:
![image](https://user-images.githubusercontent.com/6021109/117333609-f5885700-ae4d-11eb-8898-53a99064dbe4.png)

AmyL219 commented on AzureCalling/Controllers/BottomDrawerViewController.swift:
```diff
@@ -0,0 +1,62 @@
+
+import UIKit
+
+protocol BottomDrawerViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
Protocol is usually defined for an interface but not the some generation method. It doesn't contain the implement of the things. I dont think it is the best choice to define BottomDrawerViewController as a protocol even it work here. At least we should define it as a base viewcontroller, and participant list and audio selection drawer can inherit from it. 
Ideally, could the bottomDrawerViewController be the only drawer, and we input different datasource of BottomDrawerItem into the drawer? Maybe the bottomDrawerViewController has 2 different style type, one is for AudioSelection, one is for ParticipantList to build the view difference

cheukchuen committed:
Increase the length of display name in participant list

AmyL219 commented on AzureCalling/Controllers/ParticipantListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+
+    private var participantList: [BottomDrawerItem] = [BottomDrawerItem]()
+    var participantTable: UITableView!
+    var callingContext: CallingContext!
```
I don't think we need to inject the whole callingContext into this class. We just need the remote participant list

Gemakk replied:
We also need our own name and mute status.  Once we implement the live participant list, this will expand to an observer to our own mute status as well and possibly our own speaking indicator.

cheukchuen replied:
Defined a struct `ParticipantListInfo` to carry the necessary info to build the participant list, no longer rely on the entire `callingContext`

cheukchuen committed:
Define BottomDrawer as a single table

- AudioDeviceSelection and ParticipantList are both separate datasources and delegates to be supplied to BottomDrawer

AmyL219 reviewed (commented):
I think we're in the right page. Just some comments for refactoring suggestion

cheukchuen commented on AzureCalling/Controllers/AudioDeviceSelectionViewController.swift:
```diff
@@ -5,85 +5,13 @@
 import UIKit
 
-class AudioDeviceSelectionViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
+class AudioDeviceSelectionViewController: NSObject, UITableViewDelegate, UITableViewDataSource {
```
Good point. It is also a Delegate though, but I'm fine with DataSource. Where do we place these files though since they are not controllers anymore?

AmyL219 replied:
I think we could create a new sub folder named `datasource` inside `Controllers`

cheukchuen commented on AzureCalling/Controllers/BottomDrawerViewController.swift:
```diff
@@ -0,0 +1,92 @@
+
+    weak var tableViewDelegate: UITableViewDelegate?
+    var tableViewDataSource: UITableViewDataSource?
+    var tableView: UITableView!
```
Agree with the initialization, just that the properties still need to be optional since we also need to add the following init:
```swift
required init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
}
```

cheukchuen committed:
Code refactoring

cheukchuen commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -134,9 +135,23 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
+        audioDeviceSelectionViewController.createAudioDeviceOptions()
+        bottomDrawerViewController.tableViewDataSource = audioDeviceSelectionViewController
+        bottomDrawerViewController.tableViewDelegate = audioDeviceSelectionViewController
+        bottomDrawerViewController.modalPresentationStyle = .overCurrentContext
```
It works!

cheukchuen commented on AzureCalling/Controllers/ParticipantListViewController.swift:
```diff
@@ -0,0 +1,49 @@
+struct ParticipantListInfo {
+    let localDisplayName: String
+    let localIsMuted: Bool?
+    let remoteParticipants: MappedSequence<String, RemoteParticipant>
```
Replaced with `[RemoteParticipantInfo]` instead

cheukchuen committed:
Merge branch 'main' into csiow/static-participant-list

AmyL219 commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -134,9 +135,31 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
+        let participantListInfo = ParticipantListInfo(
+            localDisplayName: callingContext.displayName,
+            localIsMuted: callingContext.callIsMuted,
+            remoteParticipants: callingContext.remoteParticipants.map { (remoteParticipant) -> RemoteParticipantInfo in
```
It would be clear if the converting remoteParticipant and ListInfo are separated. 
```let remoteParticipantInfoList = callingContext.remoteParticipants.map .....
let participantListInfo = ParticipantListInfo(
            localDisplayName: callingContext.displayName,
            localIsMuted: callingContext.callIsMuted,
            remoteParticipants:remoteParticipantInfoList

AmyL219 commented on AzureCalling.xcodeproj/project.pbxproj:
```diff
@@ -158,13 +162,14 @@
 			isa = PBXGroup;
 			children = (
+				989F61952649D72F000C5668 /* datasource */,
 				1F7CD3EB25C1E9E80081D8F1 /* IntroViewController.swift */,
```
Let's use upper case `DataSource` to keep consistent with other groups

AmyL219 reviewed (commented):
Two minor comments. Others looks good to me

cheukchuen committed:
Minor code refactoring

waynemo commented on AzureCalling/Controllers/datasource/ParticipantListDataSource.swift:
```diff
@@ -0,0 +1,53 @@
+
+import UIKit
+
+struct ParticipantListInfo {
+    let localDisplayName: String
+    let localIsMuted: Bool?
+    let remoteParticipants: [RemoteParticipantInfo]
+}
+
+struct RemoteParticipantInfo {
+    let displayName: String
+    let isMuted: Bool
+}
```
Couldn't we just have a list of `ParticipantInfo` that can be used for both local and remote?

cheukchuen replied:
Good point! That should simplify and generalize the participant list

waynemo commented on AzureCalling/External/Calling/CallingContext.swift:
```diff
@@ -37,6 +37,10 @@ class CallingContext: NSObject {
         return remoteParticipantCount + 1
     }
 
+    var callIsMuted: Bool? {
```
I think we should use "is" as the prefix -> `isCallMuted`

cheukchuen committed:
Simplify participant info list to include both local and remote participants

Gemakk reviewed (approved):
Tested
